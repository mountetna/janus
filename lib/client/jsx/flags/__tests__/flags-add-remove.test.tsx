import React from 'react';
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import {render, fireEvent, waitFor, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import AddRemove from '../flags-add-remove';
import {flagsSpecWrapper} from '../../spec/flags-helpers';

const mockUsers = [
  {
    email: 'janus@two-faces.org',
    name: 'Janus Bifrons',
    projects: ['doors', 'portals'],
    flags: ['inside', 'outside']
  },
  {
    email: 'portunus@two-faces.org',
    name: 'Portunus',
    projects: ['doors'],
    flags: ['underground']
  }
];

const mockProjects = [
  {project_name: 'doors', project_name_full: 'Oak doors'},
  {project_name: 'portals', project_name_full: 'Magic portals'}
];

const mockState = {
  projects: mockProjects,
  users: mockUsers
};

const handlers = [
  rest.post('/flag_user', (req, res, ctx) => {
    return res(ctx.json({}));
  }),
  rest.get('/users', (req, res, ctx) => {
    return res(ctx.json({users: mockUsers}));
  })
];

const server = setupServer(...handlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('AddRemove', () => {
  it('renders correctly', async () => {
    let {asFragment} = render(
      <AddRemove selectedUsers={mockUsers.slice(0, 1)}/>,
      {wrapper: flagsSpecWrapper(mockState)}
    );

    await waitFor(() => screen.getByText(/1 user/));

    expect(screen.queryByText('Add Flag')).toBeTruthy();
    expect(screen.queryByText('Remove Flag')).toBeTruthy();

    expect(asFragment()).toMatchSnapshot();
  });

  it('appends flags to existing flags', async () => {
    let {asFragment} = render(
      <AddRemove selectedUsers={mockUsers.slice(0, 1)} />,
      {wrapper: flagsSpecWrapper(mockState)}
    );

    await waitFor(() => screen.getByText(/1 user/));

    fireEvent.change(screen.getByRole('textbox', {name: ''}), {
      target: {value: 'newflag'}
    });
    fireEvent.click(screen.getByText('Add Flag'));

    await waitFor(() => {
      expect(screen.queryByText(/Saved/)).toBeTruthy();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('removes flags from users', async () => {
    let {asFragment} = render(
      <AddRemove selectedUsers={mockUsers.slice(0, 1)} />,
      {wrapper: flagsSpecWrapper(mockState)}
    );

    await waitFor(() => screen.getByText(/1 user/));

    fireEvent.change(screen.getByRole('textbox', {name: ''}), {
      target: {value: 'newflag'}
    });
    fireEvent.click(screen.getByText('Remove Flag'));

    await waitFor(() => {
      expect(screen.queryByText(/Saved/)).toBeTruthy();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('reports errors to the user', async () => {
    server.use(
      rest.post('/flag_user', (req, res, ctx) => {
        // Respond with "422 Error" status for this test.
        return res(
          ctx.status(422),
          ctx.json({error: 'Must be an array of strings'})
        );
      })
    );

    let {asFragment} = render(
      <AddRemove selectedUsers={mockUsers.slice(0, 1)} />,
      {wrapper: flagsSpecWrapper(mockState)}
    );

    await waitFor(() => screen.getByText(/1 user/));

    fireEvent.change(screen.getByRole('textbox', {name: ''}), {
      target: {value: 'bad-flag'}
    });
    fireEvent.click(screen.getByText('Add Flag'));

    await waitFor(() => {
      expect(screen.queryByText(/Must be an array of strings/)).toBeTruthy();

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
