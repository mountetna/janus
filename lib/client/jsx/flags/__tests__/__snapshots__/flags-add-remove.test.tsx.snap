// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRemove appends flags to existing flags 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Add or remove a flag
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          on 1 user
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Flag text
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled"
            >
              <span>
                Flag text
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained"
        >
          Saved!
        </p>
      </div>
    </div>
    <div
      class="MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14v2z"
              />
            </svg>
          </span>
          Remove Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddRemove removes flags from users 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Add or remove a flag
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          on 1 user
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Flag text
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled"
            >
              <span>
                Flag text
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained"
        >
          Saved!
        </p>
      </div>
    </div>
    <div
      class="MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14v2z"
              />
            </svg>
          </span>
          Remove Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddRemove renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Add or remove a flag
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          on 1 user
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Flag text
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled"
            >
              <span>
                Flag text
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14v2z"
              />
            </svg>
          </span>
          Remove Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddRemove reports errors to the user 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Add or remove a flag
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          on 1 user
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-error MuiInputLabel-error MuiFormLabel-filled"
          data-shrink="true"
        >
          Flag text
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-error MuiOutlinedInput-error MuiInputBase-formControl"
        >
          <input
            aria-invalid="true"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value="bad-flag"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
            >
              <span>
                Flag text
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-error MuiFormHelperText-filled"
        >
          Must be an array of strings
        </p>
      </div>
    </div>
    <div
      class="MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button MuiButton-containedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14v2z"
              />
            </svg>
          </span>
          Remove Flag
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
